src/ucs/datastruct/hlist.h

```diff
@@ -266,6 +266,22 @@ ucs_hlist_extract_head(ucs_hlist_head_t *head)
          _elem = ucs_hlist_extract_head_elem(_head, typeof(*(_elem)), _member))
 
 
+/**
+ * Iterate over detached-head list, while removing the head element passes
+ * @a _cond or the list becomes empty.
+ *
+ * @param _elem     Variable to hold the current list element
+ * @param _head     Pointer to list head.
+ * @param _member   List element inside the containing structure.
+ * @param _cond     Condition to test for @a _head element before extract it.
+ */
+#define ucs_hlist_for_each_extract_if(_elem, _head, _member, _cond) \
+    for (_elem = ucs_hlist_head_elem(_head, typeof(*(_elem)), _member); \
+         _elem != UCS_PTR_BYTE_OFFSET(NULL, -ucs_offsetof(typeof(*(_elem)), _member)) && (_cond); \
+         _elem = ucs_hlist_extract_head_elem(_head, typeof(*(_elem)), _member), \
+         _elem = ucs_hlist_head_elem(_head, typeof(*(_elem)), _member)) \
```

Comment:
why setting _elem to head again?

Reply:
1st assignment extracts head, 2nd assign elem for next iteration, pls note, this loop extracts head in the *end* of iteration, not in *begin* like origin for_each impl.

Reply:
so i think we can replace
`_elem = ucs_hlist_extract_head_elem(_head, typeof(*(_elem)), _member)`
by just
`ucs_hlist_extract_head(_head)`
since _elem is changed again in the next line (282)